### PR TITLE
✨ Initial implementation of IndexedDataset

### PR DESCRIPTION
**Pull request by Koncopd**
subj
in progress

### CODE REVIEW

File: lamindb/dev/dataloader/dataset.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+from typing import List, Optional, Union
+
+import numpy as np
+from lamindb_setup.dev.upath import UPath
+
+from ..storage._backed_access import ArrayTypes, GroupTypes, StorageType, registry
+
+
+# this is based on sCimilarity (https://github.com/Genentech/scimilarity) Datasets
+class ListDataset:
```
Comment (falexwolf) [inline lamindb/dev/dataloader/dataset.py:12]: How about calling this `IndexedDataset`?

File: lamindb/_dataset.py [hunk 2]
```diff
@@ -329,6 +332,23 @@ def from_files(files: Iterable[File]) -> Tuple[str, Dict[str, str]]:
     return hash, feature_sets_union
 
 
+def filelist_dataset(
```
Comment (falexwolf) [inline lamindb/_dataset.py:335]: How about `indexed()`?

Comment (falexwolf) [reply lamindb/_dataset.py:335]: Or `to_indexed()`?

**Comment by falexwolf** [comment]
I'm just realizing that torch calls this Map-style dataset. But I'm almost tempted to think that that `IndexedDataset` will be clearer to users than `MappedDataset`. Or what do you think, @Koncopd?
image

**Comment by Koncopd** [comment]
Yes, you are right. Not sure, we don't expose the class anyways. I am fine with both.

**Pull request merged**